// @fluentui/web-components/dist/esm/data-grid/index.js
import { __decorate } from "tslib";
import { customElement } from '@microsoft/fast-element';
import { createDataGridCellTemplate, createDataGridRowTemplate, createDataGridTemplate, DataGrid, DataGridCell, DataGridRow, } from '@microsoft/fast-foundation';
import { DataGridStyles as gridStyles } from './data-grid.styles';
import { DataGridRowStyles as rowStyles } from './data-grid-row.styles';
import { DataGridCellStyles as cellStyles } from './data-grid-cell.styles';
const cellTemplate = createDataGridCellTemplate('fluent');
const gridTemplate = createDataGridTemplate('fluent');
const rowTemplate = createDataGridRowTemplate('fluent');
/**
 * The Fluent Data Grid Element.
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-data-grid\>
 */
let FluentDataGrid = class FluentDataGrid extends DataGrid {
};
FluentDataGrid = __decorate([
    customElement({
        name: 'fluent-data-grid',
        template: gridTemplate,
        styles: gridStyles,
    })
], FluentDataGrid);
export { FluentDataGrid };
/**
 * Styles for DataGrid
 * @public
 */
export const DataGridStyles = gridStyles;
/**
 * The Fluent Data Grid Row Element.
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-data-grid-row\>
 */
let FluentDataGridRow = class FluentDataGridRow extends DataGridRow {
};
FluentDataGridRow = __decorate([
    customElement({
        name: 'fluent-data-grid-row',
        template: rowTemplate,
        styles: rowStyles,
    })
], FluentDataGridRow);
export { FluentDataGridRow };
/**
 * Styles for DataGrid row
 * @public
 */
export const DataGridRowStyles = rowStyles;
/**
 * The Fluent Data Grid Cell Element.
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-data-grid-cell\>
 */
let FluentDataGridCell = class FluentDataGridCell extends DataGridCell {
};
FluentDataGridCell = __decorate([
    customElement({
        name: 'fluent-data-grid-cell',
        template: cellTemplate,
        styles: cellStyles,
    })
], FluentDataGridCell);
export { FluentDataGridCell };
/**
 * Styles for DataGrid cell
 * @public
 */
export const DataGridCellStyles = cellStyles;
