// @fluentui/web-components/dist/esm/horizontal-scroll/index.js
import { __decorate } from "tslib";
import { customElement } from '@microsoft/fast-element';
import { HorizontalScroll, HorizontalScrollTemplate as template } from '@microsoft/fast-foundation';
import { ActionsStyles, HorizontalScrollStyles as styles } from './horizontal-scroll.styles';
/**
 * The Fluent Horizontal Scroll Element. Implements {@link @microsoft/fast-foundation#HorizontalScrol},
 * {@link @microsoft/fast-foundation#HorizontalScrollTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-horizontal-scroll\>
 */
let FluentHorizontalScroll = class FluentHorizontalScroll extends HorizontalScroll {
    /**
     * @public
     */
    connectedCallback() {
        super.connectedCallback();
        if (this.view !== 'mobile') {
            this.$fastController.addStyles(ActionsStyles);
        }
    }
};
FluentHorizontalScroll = __decorate([
    customElement({
        name: 'fluent-horizontal-scroll',
        template,
        styles,
        shadowOptions: {
            mode: 'closed',
        },
    })
], FluentHorizontalScroll);
export { FluentHorizontalScroll };
/**
 * Styles for Horizontal Scroll
 * @public
 */
export const HorizontalScrollStyles = styles;
