// @fluentui/web-components/dist/esm/color/neutral-outline-contrast.js
import { subtract } from '../utilities/math';
import { neutralOutlineActiveDelta, neutralOutlineFocusDelta, neutralOutlineHoverDelta, neutralOutlineRestDelta, neutralPalette, } from '../fluent-design-system';
import { colorRecipeFactory, swatchFamilyToSwatchRecipeFactory, SwatchFamilyType, } from './common';
import { accessibleAlgorithm } from './accessible-recipe';
/**
 * @internal
 */
export const neutralOutlineContrastAlgorithm = colorRecipeFactory(accessibleAlgorithm(neutralPalette, 3, 0, subtract(neutralOutlineHoverDelta, neutralOutlineRestDelta), subtract(neutralOutlineActiveDelta, neutralOutlineRestDelta), subtract(neutralOutlineFocusDelta, neutralOutlineRestDelta)));
/**
 * @internal
 */
export const neutralOutlineContrast = colorRecipeFactory(neutralOutlineContrastAlgorithm);
/**
 * @internal
 */
export const neutralOutlineContrastRest = swatchFamilyToSwatchRecipeFactory(SwatchFamilyType.rest, neutralOutlineContrast);
/**
 * @internal
 */
export const neutralOutlineContrastHover = swatchFamilyToSwatchRecipeFactory(SwatchFamilyType.hover, neutralOutlineContrast);
/**
 * @internal
 */
export const neutralOutlineContrastActive = swatchFamilyToSwatchRecipeFactory(SwatchFamilyType.active, neutralOutlineContrast);
/**
 * @internal
 */
export const neutralOutlineContrastFocus = swatchFamilyToSwatchRecipeFactory(SwatchFamilyType.focus, neutralOutlineContrast);
