// @microsoft/mgt-msal-provider/dist/es6/mgt-msal-provider.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { customElement, property } from 'lit-element';
import { Providers, LoginType, MgtBaseProvider } from '@microsoft/mgt-element';
import { MsalProvider } from './MsalProvider';
/**
 * Authentication Library Provider for Microsoft personal accounts
 *
 * @export
 * @class MgtMsalProvider
 * @extends {MgtBaseProvider}
 */
let MgtMsalProvider = class MgtMsalProvider extends MgtBaseProvider {
    constructor() {
        super(...arguments);
        /**
         * String alphanumerical value relation to a specific user
         *
         * @memberof MgtMsalProvider
         */
        this.clientId = '';
    }
    /**
     * Gets whether this provider can be used in this environment
     *
     * @readonly
     * @memberof MgtMsalProvider
     */
    get isAvailable() {
        return true;
    }
    /**
     * method called to initialize the provider. Each derived class should provide their own implementation.
     *
     * @protected
     * @memberof MgtMsalProvider
     */
    initializeProvider() {
        if (this.clientId) {
            const config = {
                clientId: this.clientId,
                domainHint: this.domainHint,
                prompt: this.prompt
            };
            if (this.loginType && this.loginType.length > 1) {
                let loginType = this.loginType.toLowerCase();
                loginType = loginType[0].toUpperCase() + loginType.slice(1);
                const loginTypeEnum = LoginType[loginType];
                config.loginType = loginTypeEnum;
            }
            if (this.authority) {
                config.authority = this.authority;
            }
            if (this.scopes) {
                const scope = this.scopes.split(',');
                if (scope && scope.length > 0) {
                    config.scopes = scope;
                }
            }
            if (this.redirectUri) {
                config.redirectUri = this.redirectUri;
            }
            this.provider = new MsalProvider(config);
            Providers.globalProvider = this.provider;
        }
    }
};
__decorate([
    property({
        attribute: 'client-id',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "clientId", void 0);
__decorate([
    property({
        attribute: 'login-type',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "loginType", void 0);
__decorate([
    property(),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "authority", void 0);
__decorate([
    property({
        attribute: 'scopes',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "scopes", void 0);
__decorate([
    property({
        attribute: 'redirect-uri',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "redirectUri", void 0);
__decorate([
    property({
        attribute: 'domain-hint',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "domainHint", void 0);
__decorate([
    property({
        attribute: 'prompt',
        type: String
    }),
    __metadata("design:type", Object)
], MgtMsalProvider.prototype, "prompt", void 0);
MgtMsalProvider = __decorate([
    customElement('mgt-msal-provider')
], MgtMsalProvider);
export { MgtMsalProvider };
