// @fluentui/web-components/dist/esm/text-field/index.js
import { __decorate } from "tslib";
import { attr, customElement } from '@microsoft/fast-element';
import { TextFieldTemplate as template, TextField } from '@microsoft/fast-foundation';
import { TextFieldStyles as styles } from './text-field.styles';
/**
 * The Fluent Text Field Custom Element. Implements {@link @microsoft/fast-foundation#TextField},
 * {@link @microsoft/fast-foundation#TextFieldTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-text-field\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FluentTextField = class FluentTextField extends TextField {
    /**
     * @internal
     */
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = 'outline';
        }
    }
};
__decorate([
    attr
], FluentTextField.prototype, "appearance", void 0);
FluentTextField = __decorate([
    customElement({
        name: 'fluent-text-field',
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
            mode: 'closed',
        },
    })
], FluentTextField);
export { FluentTextField };
/**
 * Styles for TextField
 * @public
 */
export const TextFieldStyles = styles;
