// @fluentui/web-components/dist/esm/number-field/index.js
import { __decorate } from "tslib";
import { attr, customElement } from '@microsoft/fast-element';
import { NumberField, NumberFieldTemplate as template } from '@microsoft/fast-foundation';
import { NumberFieldStyles as styles } from './number-field.styles';
/**
 * The Fluent Number Field Custom Element. Implements {@link @microsoft/fast-foundation#NumberField},
 * {@link @microsoft/fast-foundation#NumberFieldTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-number-field\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FluentNumberField = class FluentNumberField extends NumberField {
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = 'outline';
        }
    }
};
__decorate([
    attr
], FluentNumberField.prototype, "appearance", void 0);
FluentNumberField = __decorate([
    customElement({
        name: 'fluent-number-field',
        shadowOptions: {
            delegatesFocus: true,
            mode: 'closed',
        },
        styles,
        template,
    })
], FluentNumberField);
export { FluentNumberField };
/**
 * Styles for NumberField
 * @public
 */
export const NumberFieldStyles = styles;
