// @fluentui/web-components/dist/esm/text-area/index.js
import { __decorate } from "tslib";
import { attr, customElement } from '@microsoft/fast-element';
import { TextAreaTemplate as template, TextArea } from '@microsoft/fast-foundation';
import { TextAreaStyles as styles } from './text-area.styles';
/**
 * The Fluent Text Area Custom Element. Implements {@link @microsoft/fast-foundation#TextArea},
 * {@link @microsoft/fast-foundation#TextAreaTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-text-area\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FluentTextArea = class FluentTextArea extends TextArea {
    /**
     * @internal
     */
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = 'outline';
        }
    }
};
__decorate([
    attr
], FluentTextArea.prototype, "appearance", void 0);
FluentTextArea = __decorate([
    customElement({
        name: 'fluent-text-area',
        template,
        styles,
        shadowOptions: {
            mode: 'closed',
        },
    })
], FluentTextArea);
export { FluentTextArea };
/**
 * Styles for TextArea
 * @public
 */
export const TextAreaStyles = styles;
