// @fluentui/web-components/dist/esm/anchor/index.js
import { __decorate } from "tslib";
import { attr, customElement } from '@microsoft/fast-element';
import { Anchor, AnchorTemplate as template } from '@microsoft/fast-foundation';
import { AnchorStyles as styles } from './anchor.styles';
/**
 * The Fluent Anchor Element. Implements {@link @microsoft/fast-foundation#Anchor},
 * {@link @microsoft/fast-foundation#AnchorTemplate}
 *
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-anchor\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
let FluentAnchor = class FluentAnchor extends Anchor {
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = 'neutral';
        }
    }
    /**
     * Applies 'icon-only' class when there is only an SVG in the default slot
     *
     * @internal
     */
    defaultSlottedContentChanged() {
        const slottedElements = this.defaultSlottedContent.filter(x => x.nodeType === Node.ELEMENT_NODE);
        if (slottedElements.length === 1 && slottedElements[0] instanceof SVGElement) {
            this.control.classList.add('icon-only');
        }
        else {
            this.control.classList.remove('icon-only');
        }
    }
};
__decorate([
    attr
], FluentAnchor.prototype, "appearance", void 0);
FluentAnchor = __decorate([
    customElement({
        name: 'fluent-anchor',
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
            mode: 'closed',
        },
    })
], FluentAnchor);
export { FluentAnchor };
/**
 * Styles for Anchor
 * @public
 */
export const AnchorStyles = styles;
