// @fluentui/web-components/dist/esm/combobox/index.js
import { __decorate } from "tslib";
import { attr, customElement } from '@microsoft/fast-element';
import { Combobox, ComboboxTemplate as template } from '@microsoft/fast-foundation';
import { ComboboxStyles as styles } from './combobox.styles';
/**
 * The Fluent Combobox Custom Element. Implements {@link @microsoft/fast-foundation#Combobox|Combobox},
 * {@link @microsoft/fast-foundation#ComboboxTemplate|ComboboxTemplate}
 *
 * @public
 * @remarks
 * HTML Element: \<fluent-combobox\>
 *
 */
let FluentCombobox = class FluentCombobox extends Combobox {
    /**
     * @internal
     */
    appearanceChanged(oldValue, newValue) {
        if (oldValue !== newValue) {
            this.classList.add(newValue);
            this.classList.remove(oldValue);
        }
    }
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = 'outline';
        }
    }
};
__decorate([
    attr({ mode: 'fromView' })
], FluentCombobox.prototype, "appearance", void 0);
FluentCombobox = __decorate([
    customElement({
        name: 'fluent-combobox',
        template,
        styles,
        shadowOptions: {
            delegatesFocus: true,
        },
    })
], FluentCombobox);
export { FluentCombobox };
/**
 * Styles for combobox
 * @public
 */
export const ComboboxStyles = styles;
